src/strings/bstring.rs

```diff
@@ -1,21 +1,54 @@
 use crate::*;
+use std::convert::TryFrom;
+use std::result::Result as StdResult;
 
-#[repr(C)]
+/// A Basic string, sometimes called a [BSTR](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/automat/bstr),
+/// is a string data type used by certain COM interfaces and interop functions.
+#[repr(transparent)]
 pub struct BString(RawPtr);
 
 impl BString {
-    pub fn new() -> BString {
+    /// Create an empty `BString`.
+    ///
+    /// This function does no allocation.
+    pub fn new() -> Self {
         Self(std::ptr::null_mut())
     }
 
+    /// Returns `true` if the string is empty.
     pub fn is_empty(&self) -> bool {
         self.len() == 0
     }
 
+    /// Returns the length of `self`.
     pub fn len(&self) -> usize {
         unsafe { SysStringLen(self.0) as usize }
     }
 
+    /// Get the string as 16-bit wide characters (wchars).
+    pub fn as_wide(&self) -> &[u16] {
+        if self.is_empty() {
+            return &[];
+        }
+
+        unsafe { std::slice::from_raw_parts(self.0 as *const u16, self.len()) }
+    }
+
+    /// Create a `BString` from a slice of 16 bit characters (wchars).
+    pub fn from_wide(value: &[u16]) -> Self {
+        if value.len() == 0 {
+            return Self::new();
+        }
+
+        unsafe { Self(SysAllocStringLen(value.as_ptr(), value.len() as u32)) }
+    }
+
+    /// Get the contents of this `BString` as a `String` lossily.
+    pub fn to_string_lossy(&self) -> String {
+        String::from_utf16_lossy(self.as_wide())
+    }
+
+    /// Frees the memory occupied by the string.
     pub fn clear(&mut self) {
         unsafe {
             SysFreeString(self.0);
```

Comment:
Nit: It might be "more correct" to store the pointer into a local, set the `self.0` to null and then call `SysFreeString` on the local. 

`SysFreeString` can't kick off a stack unwind can it?